domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -143,15 +203,247 @@ class TopicListController @Inject constructor(
       .build()
   }
 
+  private fun createUpcomingTopicSummaryFromJson(
+    topicId: String,
+    jsonObject: JSONObject
+  ): UpcomingTopic {
+    var totalChapterCount = 0
+    val storyData = jsonObject.getJSONArray("canonical_story_dicts")
+    for (i in 0 until storyData.length()) {
+      totalChapterCount += storyData
+        .getJSONObject(i)
+        .getJSONArray("node_titles")
+        .length()
+    }
+    return UpcomingTopic.newBuilder()
+      .setTopicId(topicId)
+      .setName(jsonObject.getString("topic_name"))
+      .setVersion(jsonObject.optInt("version"))
+      .setEstimatedReleaseUnixTimestamp(Date().time)
+      .setLessonThumbnail(createTopicThumbnail(jsonObject))
+      .build()
+  }
+
   private fun createOngoingStoryListFromProgress(
```

Comment:
This function is getting quite long--can you please break it into smaller, more isolated chunks by moving unrelated pieces into helper methods?

Reply:
Broken into chunks.

Reply:
You can break this further by creating separate functions for where clause.

something like:
```
        when {
            recentlyPlayerChapterProgress != null -> {
              createOngoingStoryListBasedOnRecentlyPlayed()
            lastCompletedChapterProgress != null &&
              lastCompletedChapterProgress.explorationId !=
              story.chapterList.last().explorationId ->
              createOngoingStoryListBasedOnLastCompleted()
          }
```

Reply:
Looks better. Thanks.

Reply:
Much better--thanks!